### PR TITLE
[Bug] AdamW optimizer fails to converge with default lr0=0.01

### PR DESCRIPTION
**Pull request by S-kewen**
Hi, there,

### Search before asking

- [x] I searched Ultralytics' issue reporting page and found that many people have encountered this problem.
https://github.com/ultralytics/ultralytics/issues/23232

### Ultralytics YOLO Component

Training

### Bug

When manually specifying `optimizer='AdamW'`, training fails to converge because the default learning rate `lr0=0.01` is too high for Adam-family optimizers.

**Expected behavior:**
Adam/AdamW typically requires `lr=0.001` (1e-3), while SGD uses `lr=0.01` (1e-2). The code should auto-adjust the learning rate when using Adam-family optimizers.

**Actual behavior:**
Training with AdamW uses `lr=0.01`, causing the model to fail to fit. After 270+ epochs, mAP50-95 remains below 0.1.

### Environment
- Ultralytics version: 8.4.2 (latest)

### Minimal Reproducible Example

from ultralytics import YOLO

model = YOLO('yolo11n.yaml')
model.train(
    data='dataset.yaml',
    epochs=100,
    optimizer='AdamW',
)

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Tweaks optimizer setup to auto-lower overly high learning rates for Adam-family optimizers to improve convergence 🔧📉

### 📊 Key Changes
-Adds an auto-adjustment step in `build_optimizer()` for Adam/AdamW/NAdam/RAdam/Adamax when `lr >= 0.01`.
-Logs an informational message when `lr` is reduced (e.g., `0.01 → 0.001`) to match typical Adam-family defaults.
-Forces `warmup_bias_lr` to `0.0` in this case to avoid excessively high bias warmup with Adam-family optimizers.

### 🎯 Purpose & Impact
-Prevents common training failures when users select AdamW but leave `lr0=0.01` (SGD-like default) ✅
-Improves out-of-the-box stability and convergence for Adam-family optimizers, reducing user friction 🧠
-Changes behavior for existing runs that intentionally used `lr >= 0.01` with Adam-family optimizers (now auto-reduced), so it may affect advanced/custom training setups ⚠️

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @S-kewen, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist (this is an automated message—an engineer will assist soon) 🛠️✅:

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -969,7 +969,18 @@ def build_optimizer(self, model, name="auto", lr=0.001, momentum=0.9, decay=1e-5
 
         optimizers = {"Adam", "Adamax", "AdamW", "NAdam", "RAdam", "RMSProp", "SGD", "MuSGD", "auto"}
         name = {x.lower(): x for x in optimizers}.get(name.lower())
+
         if name in {"Adam", "Adamax", "AdamW", "NAdam", "RAdam"}:
+            # Adam/AdamW typically needs lr=0.001, while SGD uses lr=0.01
+            # This prevents training failure when user specifies AdamW but forgets to adjust lr0
+            if lr >= 0.01:
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:1096]: 💡 **MEDIUM**: `if lr >= 0.01:` only reduces once (`/ 10`). For extreme user inputs (e.g. `lr=0.1`), this still leaves `0.01`, which is likely still too high for Adam-family and defeats the stated goal of preventing common failures. Consider clamping to a max Adam-family LR (e.g. `min(lr, 0.001)`) or iteratively reducing until below threshold.

File: ultralytics/engine/trainer.py [hunk 2]
```diff
@@ -969,7 +969,18 @@ def build_optimizer(self, model, name="auto", lr=0.001, momentum=0.9, decay=1e-5
...
+                lr_adjusted = lr / 10  # 0.01 -> 0.001
+                LOGGER.info(
+                    f"{colorstr('optimizer:')} Auto-adjusting lr from {lr} to {lr_adjusted} for {name} optimizer "
+                    f"(Adam-family optimizers typically need lr=1e-3, not 1e-2)"
+                )
+                lr = lr_adjusted
+                self.args.warmup_bias_lr = 0.0  # no higher than 0.01 for Adam
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:1103]: 💡 **MEDIUM**: Mutating `self.args.warmup_bias_lr` inside `build_optimizer()` can have unintended side effects outside optimizer construction (e.g., logging/config dumps, checkpoint/resume behavior, or later scheduler/warmup code that expects the user-specified value). If the intent is only to affect warmup behavior for this run, it’s safer to use a local variable (or pass an override into the warmup logic) rather than rewriting `self.args`.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good intent and a pragmatic safeguard for a common AdamW misconfiguration. Main concerns are (1) the one-step `/10` adjustment may still leave an overly high LR for larger user values, and (2) rewriting `self.args.warmup_bias_lr` inside optimizer construction may introduce unintended global side effects. Minor nit: the warmup-bias comment doesn’t match the actual forced value.

💬 Posted 2 inline comments

**Comment by glenn-jocher** [comment]
Thanks for digging into the AdamW convergence pain point and proposing a fix, S-kewen. The concern is real, but silently overriding an explicitly set lr0 (and forcing warmup_bias_lr to 0) changes behavior for everyone who manually selects AdamW, including advanced users who intentionally chose lr0=0.01. The recommended path here is optimizer='auto', which already fits an appropriate Adam-family learning rate automatically. We're going to close this one, but we'd genuinely welcome future contributions from you.
